plugins/com.google.cloud.tools.eclipse.preferences/src/com/google/cloud/tools/eclipse/preferences/areas/AreaBasedPreferencePage.java

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright ${year} Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.eclipse.preferences.areas;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IConfigurationElement;
+import org.eclipse.core.runtime.IExecutableExtension;
+import org.eclipse.core.runtime.IExtensionRegistry;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.RegistryFactory;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.jface.layout.GridLayoutFactory;
+import org.eclipse.jface.preference.FieldEditor;
+import org.eclipse.jface.preference.IPersistentPreferenceStore;
+import org.eclipse.jface.preference.IPreferenceStore;
+import org.eclipse.jface.preference.PreferencePage;
+import org.eclipse.jface.util.IPropertyChangeListener;
+import org.eclipse.jface.util.PropertyChangeEvent;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Control;
+import org.eclipse.swt.widgets.Group;
+import org.eclipse.ui.IWorkbench;
+import org.eclipse.ui.IWorkbenchPreferencePage;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * A host preference page that hosts multiple <em>preference areas</em>. These
+ * {@link PreferenceArea} are like embedded preference pages, like more complex JFace
+ * {@linkplain FieldEditor}s.
+ */
+public class AreaBasedPreferencePage extends PreferencePage
+    implements IWorkbenchPreferencePage, IExecutableExtension {
+  /**
+   * Responsible for ordering preference areas by their rank
+   */
+  public static class AreaOrdering implements Comparator<PreferenceArea> {
+    @Override
+    public int compare(PreferenceArea o1, PreferenceArea o2) {
+      return o1.getRank() - o2.getRank();
+    }
+  }
+
+  /**
+   * Builds the contents from an extension point. Could consider making this a standalone
```

Comment:
It's an interesting point.  I considered extracting this to its own class that implements `IExtensionFactory` — the Extension Registry will then call into that factory to produce the created object.  But then the developer wouldn't be able to subclass and change the pref page (if required).